README.md
# Avalanche

[ ![Build Status] [travis-image] ] [travis] [ ![Release] [release-image] ] [releases] [ ![License] [license-image] ] [license]

Avalanche is designed to mimic the behaviour of sudden increases and decreases in requests to the Snowplow Pipeline and the ability for the pipeline to scale up/down to match these changes.  This should assert that under any amount of load the pipeline will scale to compensate while not losing any data.

## Quick Start

Assuming git, **[Vagrant] [vagrant-install]** and **[VirtualBox] [virtualbox-install]** installed:

```bash
 host$ git clone https://github.com/snowplow/avalanche.git
 host$ cd avalanche
 host$ vagrant up && vagrant ssh
guest$ cd /vagrant
```

### To run a simulation from your local machine:

* Set your collector URL as an environment variable:

```bash
guest$ export SP_COLLECTOR_URL={{ the collector URL }}
guest$ export SP_SIM_TIME={{ the time to run the simulation for in minutes }}
guest$ export SP_BASELINE_USERS={{ the amount of users as a baseline }}
guest$ export SP_PEAK_USERS={{ the peak amount of users }}
```

__NOTES__:

* The baseline users will send events for the entirety of the simulation.
  - Each user will emit ~ 1 event per second.
* The peak users is the count of users at a maximal point.

* List all of the simulations available:

```bash
guest$ /vagrant/dist/gatling-charts-highcharts-bundle-2.2.0-SNAPSHOT/bin/gatling.sh -sf /vagrant/src/ -rf /vagrant/dist/results/ -m
```

* Select the simulation you wish to run.
* Wait for simulation to complete

__NOTE__: It is advised to run heavy simulations from a server in the cloud rather than from your local machine:

1. The connection of an EC2 instance will no doubt be faster than your host machine; yielding better results.
2. You will quickly saturate your home network and be unable to use your network.

The Vagrant environment should be used solely for testing purposes or for very light simulations.

### To run a simulation from an EC2 instance:

#### 1. Create a Security Group

In the EC2 Console UI select `Security Groups` from the panel on the left.

Select the `Create Security Group` button and fill in the name, description and what VPC you want to attach it to.

You will then need to add the following InBound rules with either `0.0.0./0` as the source or something more restrictive:

* SSH: `SSH | Port Range (22)`

For OutBound you can leave the default to allow everything out.

#### 2. Launch the instance

In the EC2 Console UI select the `Launch Instance` button then select the `Community AMIs` button.

##### 2.1 Choose AMI

In the search bar enter `snowplow-avalanche-0.1.0` to find the needed AMI then select it.

##### 2.2 Choose an Instance Type

For load testing using Gatling we will need a large instance with high network performance.  We recommend any of the `m4.large` up to `m4.4xlarge` depending on the intensity of the simulation you wish to run.

##### 2.3 Configure Instance

If you created your Security Group in a different VPC than the default you will need to select the same VPC in the Network field.

##### 2.4 Add Storage

For basic testing and debugging 8gb should suffice.

We also recommend changing the `Volume Type` to GP2 from Magnetic for a faster experience.

##### 2.5 Tag Instance

Add any tags you like here.

##### 2.6 Configure Security Group

Select the Security Group you created in Step 1.

##### 2.7 Review

Press the `Launch` button and select an existing, or create a new, key-pair if you want to be able to SSH onto the box.

#### 3. Run the simulations

To run the simulations you will first need to SSH onto the instance like so:

```bash
host$ chmod 400 {{ key-pair file path }}
host$ ssh -i {{ key-pair file path }} ubuntu@{{ public DNS of instance }}
```

Set your collector URL as an environment variable:

```bash
ubuntu$ export SP_COLLECTOR_URL={{ the collector URL }}
```

Set your simulation variables:

```bash
ubuntu$ export SP_SIM_TIME={{ the time to run the simulation for in minutes }}
ubuntu$ export SP_BASELINE_USERS={{ the amount of users as a baseline }}
ubuntu$ export SP_PEAK_USERS={{ the peak amount of users }}
```

You will then need to run the following command:

```bash
ubuntu$ ./snowplow/scripts/2_run.sh
```

This will launch Gatling and allow you to select the simulation you wish to run.  There are currently two options available:

* [ExponentialPeak](https://github.com/snowplow/avalanche/blob/master/src/main/com/snowplowanalytics/avalanche/ExponentialPeak.scala#L20-L44) : Will increase load on the Collector exponentially up until the maximum users.
* [LinearPeak](https://github.com/snowplow/avalanche/blob/master/src/main/com/snowplowanalytics/avalanche/LinearPeak.scala#L20-L39) : Will increase load on a linear scale up until the maximum users.

Once the simulation is finished the results will be saved to `/home/ubuntu/snowplow/results`, to then view these results you can:

* Launch a Python server in the root of the results directory:

```bash
ubuntu$ cd /home/ubuntu/snowplow/results
ubuntu$ python -m SimpleHTTPServer 3000
```

Access the results by going to your browser and entering: `http://{{ public DNS of instance }}:3000` and then navigating to your results.

* Copy the results folder from EC2 back to your local machine using `scp`:

```bash
host$ scp -i {{ key-pair file path }} -r ubuntu@{{ public DNS of instance }}:/home/ubuntu/snowplow/results {{ local directory }}
```

Access the results by opening the `index.html` file in the results directories.

## Copyright and license

Avalanche is copyright 2016 Snowplow Analytics Ltd.

Licensed under the **[Apache License, Version 2.0] [license]** (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[vagrant-install]: http://docs.vagrantup.com/v2/installation/index.html
[virtualbox-install]: https://www.virtualbox.org/wiki/Downloads
[license]: http://www.apache.org/licenses/LICENSE-2.0

[travis]: https://travis-ci.org/snowplow/avalanche
[travis-image]: https://travis-ci.org/snowplow/avalanche.svg?branch=master

[release-image]: http://img.shields.io/badge/release-0.1.0-blue.svg?style=flat
[releases]: https://github.com/snowplow/avalanche/releases

[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat
[license]: http://www.apache.org/licenses/LICENSE-2.0


CHANGELOG
Version 0.1.0 (2016-05-17)
--------------------------
Initial release


scripts/1_provision.sh
#!/bin/bash -e

sudo apt-get update
sudo apt-get install -y unzip

#############
# Constants #
#############

main_dir=/home/ubuntu/snowplow
gatling_dir=$main_dir/gatling
results_dir=$main_dir/results
src_dir=$main_dir/src

##################
# Install Java 8 #
##################

sudo add-apt-repository ppa:webupd8team/java -y
sudo apt-get update
echo oracle-java8-installer shared/accepted-oracle-license-v1-1 select true | sudo /usr/bin/debconf-set-selections
sudo apt-get install oracle-java8-installer -y

###################
# Install Gatling #
###################

gatling_archive='gatling-charts-highcharts-bundle-2.2.1-20160421.110654-1-bundle.zip'
gatling_url='https://oss.sonatype.org/content/repositories/snapshots/io/gatling/highcharts/gatling-charts-highcharts-bundle/2.2.1-SNAPSHOT/gatling-charts-highcharts-bundle-2.2.1-20160421.110654-1-bundle.zip'

mkdir -p $gatling_dir
mkdir -p $results_dir
mkdir -p $src_dir

sudo wget $gatling_url -P $gatling_dir
sudo unzip $gatling_dir/$gatling_archive -d $gatling_dir


src/main/com/snowplowanalytics/avalanche/ExponentialPeak.scala
/*
 * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.avalanche

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

/**
 * Will create an Exponential load increase on a Snowplow Pipeline.
 *
 * Simulation Breakdown Example:
 *
 * - SP_SIM_TIME=60
 * - Runtime: 60 minutes
 *
 * - SP_BASELINE_USERS=100
 * - Baseline: ~ 6000 events/minute
 *   - Runs for 60 minutes
 *
 * - SP_PEAK_USERS=10000
 *   - 1. Add 10000/500  Users over 20 minutes ~ 20
 *   - 2. Add 10000/250  Users over 18 minutes ~ 40
 *   - 3. Add 10000/125  Users over 16 minutes ~ 80
 *   - 4. Add 10000/62   Users over 14 minutes ~ 160
 *   - 5. Add 10000/31   Users over 12 minutes ~ 320
 *   - 6. Add 10000/16   Users over 10 minutes ~ 625
 *   - 7. Add 10000/8    Users over 8 minutes  ~ 1250
 *   - 8. Add 10000/4    Users over 6 minutes  ~ 2500
 *   - 9. Add 10000/2    Users over 4 minutes  ~ 5000
 *
 * - All users will be active at the ~25 minute mark and will run concurrently for 10 minutes
 */
class ExponentialPeak extends Simulation {

  // Get ENV Vars for simulation
  val SimulationTime = sys.env("SP_SIM_TIME").toInt
  val BaselineUsers = sys.env("SP_BASELINE_USERS").toInt
  val PeakUsers = sys.env("SP_PEAK_USERS").toInt

  // Raw Events
  val PageViewEvent = "e=pv&page=DemoPageTitle&refr=DemoPageReferrer&url=DemoPageUrl&co=%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fcontexts%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%5B%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fmobile_context%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22deviceManufacturer%22%3A%22Apple%20Inc.%22%2C%22osVersion%22%3A%229.2%22%2C%22osType%22%3A%22ios%22%2C%22deviceModel%22%3A%22iPhone%22%2C%22networkType%22%3A%22wifi%22%2C%22appleIdfv%22%3A%22A229D70E-3C5F-4F75-994C-7AFC610FDCC4%22%7D%7D%2C%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fclient_session%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22previousSessionId%22%3A%2227f32674-c57c-4ad6-9a39-d43de062c748%22%2C%22firstEventId%22%3A%22ecadf00a-5fd3-4728-a305-a334353a0d24%22%2C%22sessionId%22%3A%2217d7d5dc-0ae5-44a3-af13-4f13962d57fb%22%2C%22userId%22%3A%221e6f3636-213f-41af-a7c6-cc6bb4b9455a%22%2C%22sessionIndex%22%3A7%2C%22storageMechanism%22%3A%22SQLITE%22%7D%7D%5D%7D"
  val StructuredEvent = "e=se&se_la=DemoLabel&se_pr=DemoProperty&se_ca=DemoCategory&se_va=5&se_ac=DemoAction&co=%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fcontexts%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%5B%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fmobile_context%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22deviceManufacturer%22%3A%22Apple%20Inc.%22%2C%22osVersion%22%3A%229.2%22%2C%22osType%22%3A%22ios%22%2C%22deviceModel%22%3A%22iPhone%22%2C%22networkType%22%3A%22wifi%22%2C%22appleIdfv%22%3A%22A229D70E-3C5F-4F75-994C-7AFC610FDCC4%22%7D%7D%2C%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fclient_session%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22previousSessionId%22%3A%2227f32674-c57c-4ad6-9a39-d43de062c748%22%2C%22firstEventId%22%3A%22ecadf00a-5fd3-4728-a305-a334353a0d24%22%2C%22sessionId%22%3A%2217d7d5dc-0ae5-44a3-af13-4f13962d57fb%22%2C%22userId%22%3A%221e6f3636-213f-41af-a7c6-cc6bb4b9455a%22%2C%22sessionIndex%22%3A7%2C%22storageMechanism%22%3A%22SQLITE%22%7D%7D%5D%7D"

  object Events {
    val baseline = during(SimulationTime minutes, "n") {
      exec(http("PageView Event")
        .get("/i?" + "&" + PageViewEvent))
      .pause(1 seconds)
      .exec(http("Structured Event")
        .get("/i?" + "&" + StructuredEvent))
      .pause(1 seconds)
    }
    val peak = during((SimulationTime/2) minutes, "n") {
      exec(http("PageView Event")
        .get("/i?" + "&" + PageViewEvent))
      .pause(1 seconds)
      .exec(http("Structured Event")
        .get("/i?" + "&" + StructuredEvent))
      .pause(1 seconds)
    }
  }

  // HTTP Configuration
  val httpConf = http
    .baseURL(sys.env("SP_COLLECTOR_URL"))
    .acceptHeader("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
    .doNotTrackHeader("1")
    .acceptLanguageHeader("en-US,en;q=0.5")
    .acceptEncodingHeader("gzip, deflate")
    .userAgentHeader("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:16.0) Gecko/20100101 Firefox/16.0")

  // Create Users
  val baseline = scenario("Baseline").exec(Events.baseline)
  val peak1 = scenario("Peak 1").exec(Events.peak)
  val peak2 = scenario("Peak 2").exec(Events.peak)
  val peak3 = scenario("Peak 3").exec(Events.peak)
  val peak4 = scenario("Peak 4").exec(Events.peak)
  val peak5 = scenario("Peak 5").exec(Events.peak)
  val peak6 = scenario("Peak 6").exec(Events.peak)
  val peak7 = scenario("Peak 7").exec(Events.peak)
  val peak8 = scenario("Peak 8").exec(Events.peak)
  val peak9 = scenario("Peak 9").exec(Events.peak)

  // Divide the Sim Time into 30 parts
  val TimeSegment: Int = Utils.getPositiveCount(SimulationTime, 30)

  // Simulation
  setUp(
    baseline.inject(atOnceUsers(BaselineUsers)),
    peak1.inject(nothingFor((TimeSegment * 2)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 500)) over ((TimeSegment * 10) minutes)),
    peak2.inject(nothingFor((TimeSegment * 3)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 250)) over ((TimeSegment * 9)  minutes)),
    peak3.inject(nothingFor((TimeSegment * 4)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 125)) over ((TimeSegment * 8)  minutes)),
    peak4.inject(nothingFor((TimeSegment * 5)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 62))  over ((TimeSegment * 7)  minutes)),
    peak5.inject(nothingFor((TimeSegment * 6)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 31))  over ((TimeSegment * 6)  minutes)),
    peak6.inject(nothingFor((TimeSegment * 7)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 16))  over ((TimeSegment * 5)  minutes)),
    peak7.inject(nothingFor((TimeSegment * 8)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 8))   over ((TimeSegment * 4)  minutes)),
    peak8.inject(nothingFor((TimeSegment * 9)  minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 4))   over ((TimeSegment * 3)  minutes)),
    peak9.inject(nothingFor((TimeSegment * 10) minutes), rampUsers(Utils.getPositiveCount(PeakUsers, 2))   over ((TimeSegment * 2)  minutes))
  ).protocols(httpConf)
}

object Utils {

  /**
   * Ensures we do not get a zero value for any division.
   *
   * @param total The total to divide by
   * @param divisor The count to divide by
   * @return either the result or 1
   */
  def getPositiveCount(total: Double, divisor: Double): Int = {
    val Count: Double = (total/divisor) match {
      case c if c < 1 => 1
      case c => c
    }
    return Count.toInt
  }
}


src/main/com/snowplowanalytics/avalanche/LinearPeak.scala
/*
 * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.avalanche

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

/**
 * Will create a Linear load increase on a Snowplow Pipeline.
 *
 * Simulation Breakdown Example:
 *
 * - SP_SIM_TIME=60
 * - Runtime: 60 minutes
 *
 * - SP_BASELINE_USERS=100
 * - Baseline: ~ 6000 events/minute
 *   - Runs for 60 minutes
 *
 * - SP_PEAK_USERS=10000
 * - Peak: ~ 600000 events/minute
 *   - Starts 5 minutes after baseline
 *   - Ramps up for 20 minutes
 *   - Full peak for 10 minutes
 *   - Ramps down for 20 minutes
 *   - Ends 5 minutes before baseline
 */
class LinearPeak extends Simulation {

  // Get ENV Vars for simulation
  val SimulationTime = sys.env("SP_SIM_TIME").toInt
  val BaselineUsers = sys.env("SP_BASELINE_USERS").toInt
  val PeakUsers = sys.env("SP_PEAK_USERS").toInt

  // Raw Events
  val PageViewEvent = "e=pv&page=DemoPageTitle&refr=DemoPageReferrer&url=DemoPageUrl&co=%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fcontexts%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%5B%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fmobile_context%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22deviceManufacturer%22%3A%22Apple%20Inc.%22%2C%22osVersion%22%3A%229.2%22%2C%22osType%22%3A%22ios%22%2C%22deviceModel%22%3A%22iPhone%22%2C%22networkType%22%3A%22wifi%22%2C%22appleIdfv%22%3A%22A229D70E-3C5F-4F75-994C-7AFC610FDCC4%22%7D%7D%2C%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fclient_session%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22previousSessionId%22%3A%2227f32674-c57c-4ad6-9a39-d43de062c748%22%2C%22firstEventId%22%3A%22ecadf00a-5fd3-4728-a305-a334353a0d24%22%2C%22sessionId%22%3A%2217d7d5dc-0ae5-44a3-af13-4f13962d57fb%22%2C%22userId%22%3A%221e6f3636-213f-41af-a7c6-cc6bb4b9455a%22%2C%22sessionIndex%22%3A7%2C%22storageMechanism%22%3A%22SQLITE%22%7D%7D%5D%7D"
  val StructuredEvent = "e=se&se_la=DemoLabel&se_pr=DemoProperty&se_ca=DemoCategory&se_va=5&se_ac=DemoAction&co=%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fcontexts%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%5B%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fmobile_context%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22deviceManufacturer%22%3A%22Apple%20Inc.%22%2C%22osVersion%22%3A%229.2%22%2C%22osType%22%3A%22ios%22%2C%22deviceModel%22%3A%22iPhone%22%2C%22networkType%22%3A%22wifi%22%2C%22appleIdfv%22%3A%22A229D70E-3C5F-4F75-994C-7AFC610FDCC4%22%7D%7D%2C%7B%22schema%22%3A%22iglu%3Acom.snowplowanalytics.snowplow%5C%2Fclient_session%5C%2Fjsonschema%5C%2F1-0-1%22%2C%22data%22%3A%7B%22previousSessionId%22%3A%2227f32674-c57c-4ad6-9a39-d43de062c748%22%2C%22firstEventId%22%3A%22ecadf00a-5fd3-4728-a305-a334353a0d24%22%2C%22sessionId%22%3A%2217d7d5dc-0ae5-44a3-af13-4f13962d57fb%22%2C%22userId%22%3A%221e6f3636-213f-41af-a7c6-cc6bb4b9455a%22%2C%22sessionIndex%22%3A7%2C%22storageMechanism%22%3A%22SQLITE%22%7D%7D%5D%7D"

  object Events {
    val baseline = during(SimulationTime minutes) {
      exec(http("PageView Event")
        .get("/i?" + PageViewEvent))
      .pause(1 seconds)
      .exec(http("Structured Event")
        .get("/i?" + StructuredEvent))
      .pause(1 seconds)
    }
    val peak = during((SimulationTime/2) minutes, "i") {
      exec(http("PageView Event")
        .get("/i?" + PageViewEvent))
      .pause(1 seconds)
      .exec(http("Structured Event")
        .get("/i?" + StructuredEvent))
      .pause(1 seconds)
    }
  }

  // HTTP Configuration
  val httpConf = http
    .baseURL(sys.env("SP_COLLECTOR_URL"))
    .acceptHeader("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
    .doNotTrackHeader("1")
    .acceptLanguageHeader("en-US,en;q=0.5")
    .acceptEncodingHeader("gzip, deflate")
    .userAgentHeader("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:16.0) Gecko/20100101 Firefox/16.0")

  // Create Users
  val baseline = scenario("Baseline").exec(Events.baseline)
  val peak = scenario("Peak").exec(Events.peak)

  // Simulation
  setUp(
    baseline.inject(atOnceUsers(BaselineUsers)),
    peak.inject(nothingFor((SimulationTime/12) minutes), rampUsers(PeakUsers) over ((SimulationTime/3) minutes))
  ).protocols(httpConf)
}
